File: radar-commons-android/src/main/java/org/radarcns/android/util/NotificationHandler.java
package org.radarcns.android.util;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.media.AudioAttributes;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.RequiresApi;

import org.radarcns.android.R;
import org.radarcns.android.RadarApplication;

import static android.app.Notification.DEFAULT_SOUND;
import static android.app.Notification.DEFAULT_VIBRATE;
import static android.content.Context.NOTIFICATION_SERVICE;

/**
 * Handle notifications and notification channels.
 */
public class NotificationHandler {
    /** Notification channel ID for informational messages. No user response required. */
    public static final String NOTIFICATION_CHANNEL_INFO = "NotificationHandler.INFO";
    /**
     * Notification channel ID for tasks and active notification messages.
     * User response is requested.
     */
    public static final String NOTIFICATION_CHANNEL_NOTIFY = "NotificationHandler.NOTIFY";
    /**
     * Notification channel ID for missed tasks and app failure notification messages.
     * User response is strongly requested.
     */
    public static final String NOTIFICATION_CHANNEL_ALERT = "NotificationHandler.ALERT";
    /**
     * Notification channel ID for missed tasks and app failure notification messages.
     * User response is required.
     */
    public static final String NOTIFICATION_CHANNEL_FINAL_ALERT = "NotificationHandler.FINAL_ALERT";

    private final Context context;

    public NotificationHandler(Context context) {
        this.context = context;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createNotificationChannels();
        }
    }

    private NotificationManager manager() {
        return (NotificationManager) context.getSystemService(NOTIFICATION_SERVICE);
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void createNotificationChannels() {
        NotificationManager notificationManager = manager();

        if (notificationManager == null) {
            return;
        }

        notificationManager.createNotificationChannel(newNotificationChannel(
                NOTIFICATION_CHANNEL_INFO, NotificationManager.IMPORTANCE_LOW,
                R.string.channel_info_name, R.string.channel_info_description));

        notificationManager.createNotificationChannel(newNotificationChannel(
                NOTIFICATION_CHANNEL_NOTIFY, NotificationManager.IMPORTANCE_DEFAULT,
                R.string.channel_notify_name, R.string.channel_notify_description));

        notificationManager.createNotificationChannel(newNotificationChannel(
                NOTIFICATION_CHANNEL_ALERT, NotificationManager.IMPORTANCE_HIGH,
                R.string.channel_alert_name, R.string.channel_alert_description));

        NotificationChannel importantChannel = newNotificationChannel(
                NOTIFICATION_CHANNEL_FINAL_ALERT, NotificationManager.IMPORTANCE_HIGH,
                R.string.channel_final_alert_name, R.string.channel_final_alert_description);

        Uri alarmSound = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM);
        importantChannel.setSound(alarmSound, new AudioAttributes.Builder()
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .setUsage(AudioAttributes.USAGE_NOTIFICATION_EVENT)
                .build());

        notificationManager.createNotificationChannel(importantChannel);
    }

    /**
     * Creates a notification channel that is not yet added to the notification manager.
     * @param id channel ID
     * @param importance NotificationManager importance constant
     * @param name string resource ID of the human readable name
     * @param description string resource ID of the human readable description.
     * @return not yet added notification channel
     */
    @RequiresApi(api = Build.VERSION_CODES.O)
    public NotificationChannel newNotificationChannel(String id, int importance, int name, int description) {
        String nameString = context.getString(name);
        String descriptionString = context.getString(description);
        NotificationChannel mChannel = new NotificationChannel(id, nameString, importance);
        mChannel.setDescription(descriptionString);
        return mChannel;
    }

    public Notification.Builder builder(String channel, boolean includeIntent) {
        Notification.Builder builder;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            builder = new Notification.Builder(context, channel);
        } else {
            builder = new Notification.Builder(context);
            switch (channel) {
                case NOTIFICATION_CHANNEL_INFO:
                    builder.setPriority(Notification.PRIORITY_LOW);
                    break;
                case NOTIFICATION_CHANNEL_NOTIFY:
                    builder.setPriority(Notification.PRIORITY_DEFAULT);
                    break;
                case NOTIFICATION_CHANNEL_ALERT:
                    builder.setDefaults(DEFAULT_VIBRATE)
                            .setPriority(Notification.PRIORITY_HIGH);
                    break;
                case NOTIFICATION_CHANNEL_FINAL_ALERT:
                    builder.setDefaults(DEFAULT_SOUND | DEFAULT_VIBRATE)
                            .setPriority(Notification.PRIORITY_HIGH);
                    break;
                default:
                    // no further action
                    break;
            }
            if (channel.equals(NOTIFICATION_CHANNEL_ALERT)) {
                builder.setDefaults(DEFAULT_VIBRATE)
                        .setPriority(Notification.PRIORITY_HIGH);
            } else if (channel.equals(NOTIFICATION_CHANNEL_FINAL_ALERT)) {
                builder.setDefaults(DEFAULT_SOUND | DEFAULT_VIBRATE)
                        .setPriority(Notification.PRIORITY_HIGH);
            }
        }
        return updateNotificationAppSettings(builder, includeIntent);
    }

    public Notification.Builder updateNotificationAppSettings(Notification.Builder builder, boolean includeIntent) {
        Context applicationContext = context.getApplicationContext();
        if (applicationContext instanceof RadarApplication) {
            RadarApplication app = (RadarApplication) applicationContext;
            builder.setLargeIcon(app.getLargeIcon())
                    .setSmallIcon(app.getSmallIcon());
        }

        if (includeIntent) {
            Intent intent = context.getPackageManager().getLaunchIntentForPackage(context.getPackageName());
            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
            builder.setContentIntent(pendingIntent);
        }

        return builder.setWhen(System.currentTimeMillis());
    }
}

File: radar-commons-android/src/main/java/org/radarbase/android/util/NotificationHandler.kt
package org.radarbase.android.util

import android.app.Notification
import android.app.Notification.DEFAULT_SOUND
import android.app.Notification.DEFAULT_VIBRATE
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Context.NOTIFICATION_SERVICE
import android.media.AudioAttributes
import android.media.RingtoneManager
import android.os.Build
import androidx.annotation.RequiresApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import org.radarbase.android.R
import org.radarbase.android.RadarApplication
import org.slf4j.LoggerFactory
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Handle notifications and notification channels.
 */
class NotificationHandler(private val context: Context) {
    private val isCreated = AtomicBoolean(false)

    val manager : NotificationManager?
        get() = context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager?

    suspend fun onCreate() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            manager?.createNotificationChannels()
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private suspend fun NotificationManager.createNotificationChannels() {
        if (!isCreated.compareAndSet(false, true)) return
        logger.debug("Creating notification channels")
        coroutineScope {
            launch(Dispatchers.IO) {
                createNotificationChannel(
                    NOTIFICATION_CHANNEL_INFO,
                    NotificationManager.IMPORTANCE_LOW,
                    R.string.channel_info_name, R.string.channel_info_description
                )
            }

            launch(Dispatchers.IO) {
                createNotificationChannel(
                    NOTIFICATION_CHANNEL_NOTIFY,
                    NotificationManager.IMPORTANCE_DEFAULT,
                    R.string.channel_notify_name, R.string.channel_notify_description
                )
            }

            launch(Dispatchers.IO) {
                createNotificationChannel(
                    NOTIFICATION_CHANNEL_ALERT,
                    NotificationManager.IMPORTANCE_HIGH,
                    R.string.channel_alert_name, R.string.channel_alert_description
                )
            }

            launch(Dispatchers.IO) {
                createNotificationChannel(
                    NOTIFICATION_CHANNEL_FINAL_ALERT,
                    NotificationManager.IMPORTANCE_HIGH,
                    R.string.channel_final_alert_name,
                    R.string.channel_final_alert_description
                ) {
                    setSound(
                        RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM),
                        AudioAttributes.Builder().apply {
                            setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                            setUsage(AudioAttributes.USAGE_NOTIFICATION_EVENT)
                        }.build()
                    )
                }
            }
        }
    }

    /**
     * Creates a notification channel that is not yet added to the notification manager.
     * @param id channel ID
     * @param importance NotificationManager importance constant
     * @param name string resource ID of the human readable name
     * @param description string resource ID of the human readable description.
     * @return not yet added notification channel
     */
    @RequiresApi(api = Build.VERSION_CODES.O)
    fun NotificationManager.createNotificationChannel(id: String, importance: Int, name: Int,
                                                   description: Int,
                                                   update: (NotificationChannel.() -> Unit)? = null) {
        val channel = NotificationChannel(id, context.getString(name), importance).apply {
            setDescription(context.getString(description))
            update?.let { it() }
        }
        createNotificationChannel(channel)
    }

    /**
     * Creates a notification. Example usage:
     * ```
     * notificationHandler.create(NOTIFICATION_CHANNEL_INFO, true) {
     *     setContentText(getString(R.string.my_text))
     *     setContentTitle(getString(R.string.myTitle))
     * }
     * ```
     * @param channel channel name to send notification to
     * @param includeStartIntent whether to include an intent to start the current app when the
     *                           notification is clicked.
     * @param buildNotification notification builder function, where notification properties can be
     *                          set.
     */
    fun create(channel: String, includeStartIntent: Boolean,
               buildNotification: Notification.Builder.() -> Unit): Notification {
        val builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Notification.Builder(context, channel)
        } else {
            deprecatedBuilder(channel)
        }
        return builder.apply {
            updateNotificationAppSettings(includeStartIntent)
            buildNotification()
        }.build()
    }

    /**
     * Shows a to the user notification. Example usage:
     * ```
     * notificationHandler.notify(MY_UNIQUE_ID, NOTIFICATION_CHANNEL_INFO, true) {
     *     setContentText(getString(R.string.my_text))
     *     setContentTitle(getString(R.string.myTitle))
     * }
     * ```
     * @param id a number, unique within the app for this notification. If a notification with a
     *           duplicate ID is used, the current notification with that ID may be removed.
     * @param channel channel name to send notification to
     * @param includeStartIntent whether to include an intent to start the current app when the
     *                           notification is clicked.
     * @param buildNotification notification builder function, where notification properties can be
     *                          set.
     * @return a registration of the notification, which can be used to cancel the notification
     *         at a later time.
     */
    fun notify(id: Int, channel: String, includeStartIntent: Boolean,
               buildNotification: Notification.Builder.() -> Unit): NotificationRegistration {
        try {
            manager?.notify(id, create(channel, includeStartIntent, buildNotification))
        } catch (ex: Exception) {
            logger.error("Failed to show notification {}", id, ex)
        }
        return NotificationRegistration(manager, id)
    }

    fun cancel(notificationId: Int) {
        manager?.cancel(notificationId)
    }

    @Suppress("DEPRECATION")
    private fun deprecatedBuilder(channel: String): Notification.Builder {
        return Notification.Builder(context).apply {
            when (channel) {
                NOTIFICATION_CHANNEL_INFO -> setPriority(Notification.PRIORITY_LOW)
                NOTIFICATION_CHANNEL_NOTIFY -> setPriority(Notification.PRIORITY_DEFAULT)
                NOTIFICATION_CHANNEL_ALERT -> {
                    setDefaults(DEFAULT_VIBRATE)
                    setPriority(Notification.PRIORITY_HIGH)
                }
                NOTIFICATION_CHANNEL_FINAL_ALERT -> {
                    setDefaults(DEFAULT_SOUND or DEFAULT_VIBRATE)
                    setPriority(Notification.PRIORITY_HIGH)
                }
                else -> {}
            }// no further action
        }
    }

    private fun Notification.Builder.updateNotificationAppSettings(includeIntent: Boolean) {
        (context.applicationContext as? RadarApplication)?.let { app ->
            setLargeIcon(app.largeIcon)
            setSmallIcon(app.smallIcon)
        }

        if (includeIntent) {
            val intent = context.packageManager.getLaunchIntentForPackage(context.packageName)
            val pendingIntent = PendingIntent.getActivity(
                context,
                NOTIFICATION_REQUEST_CODE,
                intent,
                0.toPendingIntentFlag()
            )
            setContentIntent(pendingIntent)
        }

        setWhen(System.currentTimeMillis())
    }

    companion object {
        private val logger = LoggerFactory.getLogger(NotificationHandler::class.java)

        /** Notification channel ID for informational messages. No user response required.  */
        const val NOTIFICATION_CHANNEL_INFO = "org.radarbase.android.NotificationHandler.INFO"
        /**
         * Notification channel ID for tasks and active notification messages.
         * User response is requested.
         */
        const val NOTIFICATION_CHANNEL_NOTIFY = "org.radarbase.android.NotificationHandler.NOTIFY"
        /**
         * Notification channel ID for missed tasks and app failure notification messages.
         * User response is strongly requested.
         */
        const val NOTIFICATION_CHANNEL_ALERT = "org.radarbase.android.NotificationHandler.ALERT"
        /**
         * Notification channel ID for missed tasks and app failure notification messages.
         * User response is required.
         */
        const val NOTIFICATION_CHANNEL_FINAL_ALERT = "org.radarbase.android.NotificationHandler.FINAL_ALERT"

        private const val NOTIFICATION_REQUEST_CODE = 27581
    }

    data class NotificationRegistration(
            private val manager: NotificationManager?,
            private val id: Int
    ) {
        /** Cancels the notification, if possible. */
        fun cancel() {
            try {
                manager?.cancel(id)
            } catch (ex: Exception) {
                logger.error("Failed to cancel notification {}", id, ex)
            }
        }
    }
}
